otherlibs/fiber/pool.ml

```diff
@@ -20,26 +16,58 @@ let running t k =
   | Open -> k true
   | Closed -> k false
 
-let create () = { mvar = Mvar.create (); status = Open }
+let create () = { tasks = Queue.create (); runner = None; status = Open }
 
 let task t ~f k =
   match t.status with
   | Closed ->
     Code_error.raise "pool is closed. new tasks may not be submitted" []
-  | Open -> Mvar.write t.mvar (Task f) k
-
-let stream t =
-  Stream.In.create (fun () ->
-      let+ next = Mvar.read t.mvar in
-      match next with
-      | Done -> None
-      | Task task -> Some task)
+  | Open -> (
+    match t.runner with
+    | None ->
+      Queue.push t.tasks f;
+      k ()
+    | Some r ->
+      assert (Queue.is_empty t.tasks);
+      t.runner <- None;
+      Queue.push t.tasks f;
+      resume r () k)
 
 let stop t k =
   match t.status with
   | Closed -> k ()
-  | Open ->
+  | Open -> (
     t.status <- Closed;
-    Mvar.write t.mvar Done k
+    match t.runner with
+    | None -> k ()
+    | Some r ->
+      t.runner <- None;
+      resume r () k)
 
-let run t = stream t |> Stream.In.parallel_iter ~f:(fun task -> task ())
+let run t k =
+  match t.runner with
+  | Some _ ->
+    Code_error.raise "Fiber.Pool.run: concurent calls to run aren't allowed" []
+  | None ->
+    let n = ref 1 in
```

Comment:
What does this counter counts?

Reply:
(Maybe use a more informative name.)

Reply:
Will do. It's the number of fibers that are running in the pool. Only when this number reaches zero we may call the final continuation to finish `run`.